### PR TITLE
GeoZone model refactoring

### PR DESCRIPTION
**Pull request by quaxsze**
fix https://github.com/etalab/data.gouv.fr/issues/1116

### CODE REVIEW

File: udata/core/spatial/commands.py [hunk 1]
```diff
@@ -225,16 +160,15 @@ def migrate():
...
-            zone = qs(level=level, code=code).valid_at(validity).first()
+            level, code = geoids.parse(current_zone.id)
+            zone = qs(level=level, code=code).first()
             if not zone:
                 log.warning('No match for %s: skipped', current_zone.id)
                 counter['skipped'] += 1
                 continue
             previous = None
-            while not zone.is_current and len(zone.successors) == 1 and zone.id != previous:
+            while zone.id != previous:
```
Comment (maudetes) [inline udata/core/spatial/commands.py:170]: We probably don't need this while loop entirely

File: udata/core/spatial/models.py [hunk 2]
```diff
@@ -91,27 +62,14 @@ class GeoZone(db.Document):
...
-    successors = db.ListField()
-    population = db.IntField()
-    area = db.FloatField()
-    wikipedia = db.StringField()
-    wikidata = db.StringField()
-    dbpedia = db.StringField()
-    flag = db.ImageField(fs=logos)
-    blazon = db.ImageField(fs=logos)
-    logo = db.ImageField(fs=logos)
+    type = db.StringField(required=True)
```
Comment (maudetes) [inline udata/core/spatial/models.py:66]: doesn't type and level contains the same info?

File: udata/tests/features/territories/test_territories_api.py [hunk 3]
```diff
@@ -58,23 +54,11 @@ def test_suggest_town_by_insee_code(self):
         result = response.json[0]
         self.assertEqual(result['id'], self.arles.id)
         self.assertEqual(result['title'], self.arles.name)
-        self.assertEqual(result['parent'], self.bdr.name)
-
-    def test_suggest_town_by_postal_code(self):
```
Comment (maudetes) [inline udata/tests/features/territories/test_territories_api.py:63]: hmm, we're loosing this functionality, don't know if it matters

**Review by maudetes — COMMENTED** [review]
Oh my, what a clean! :broom: :clap: Thank you a lot!

I'm unsure about the migration process for users with existing GeoZones and Levels.

And don't forget the changelog (that could warn about migration) :pray:

File: udata/core/spatial/api_fields.py [hunk 4]
```diff
@@ -75,5 +67,5 @@
         description='The territory main code', required=True),
     'level': fields.String(
         description='The territory administrative level', required=True),
-    'keys': fields.Raw(description='The territory known codes')
+    'uri': fields.String(description='The zone uri', required=True)
```
[the remark this answers is not in the document]
Comment (quaxsze) [reply udata/core/spatial/api_fields.py:70]: this is not GeoLevel this is the result of Geozone suggest

**Review by maudetes — COMMENTED** [review]
Some tests made on migration

| Indicateur                                                                                                                | Before                                                                                                                                                                                                                                                                                                                                                                                                                                                             | After                                                                                                                                                                                                                                                                                                                                                                                              |
|------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Nombre de JDDs avec au moins une geozone (1)                                                                | 9781                                                                                                                                                                                                                                                                                                                                                                                                                                                               | 4669                                                                                                                                                                                                                                                                                                                                                                                               |
| geozones les plus utilisés par les JDDs (2) | [(<GeoZone: country:fr>, 126),<br> (<GeoZone: country-subset:fr:dromcom>, 71),<br> (<GeoZone: country-subset:fr:metro>, 58),<br> (<GeoZone: country-group:ue>, 21),<br> (<GeoZone: fr:commune:31555@1943-01-01>, 19),<br> (<GeoZone: country-subset:fr:drom>, 17),<br> (<GeoZone: fr:commune:31069@1943-01-01>, 16),<br> (<GeoZone: country-group:world>, 15),<br> (<GeoZone: fr:commune:31022@1943-01-01>, 15),<br> (<GeoZone: fr:commune:31032@1943-01-01>, 15)] | [(<GeoZone: fr:commune:31555>, 19),<br> (<GeoZone: fr:commune:31069>, 16),<br> (<GeoZone: fr:commune:31022>, 15),<br> (<GeoZone: fr:commune:31032>, 15),<br> (<GeoZone: fr:commune:31044>, 15),<br> (<GeoZone: fr:commune:31056>, 15),<br> (<GeoZone: fr:commune:31149>, 15),<br> (<GeoZone: fr:commune:31230>, 15),<br> (<GeoZone: fr:commune:31490>, 15),<br> (<GeoZone: fr:commune:31541>, 15)] |
| Nombre de geozones (3)                                                                                             | 51321                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 39956                                                                                                                                                                                                                                                                                                                                                                                              |
| Nombre de geozones par geolevel (4)                                                      | [('fr:commune', 43176),<br> ('fr:epci', 7468),<br> ('fr:arrondissement', 342),<br> ('country', 174),<br> ('fr:departement', 114),<br> ('fr:region', 35),<br> ('fr:collectivite', 7),<br> ('country-subset', 3),<br> ('country-group', 2)]                                                                                                                                                                                                                          | [('fr:commune', 39223),<br> ('fr:arrondissement', 390),<br> ('country', 198),<br> ('fr:departement', 111),<br> ('fr:region', 34)]                                                                                                                                                                                                                                                                  |

(1) `Dataset.objects(spatial__zones__0__exists=True).count()`
(2) `Counter(zone for dat in Dataset.objects(spatial__zones__0__exists=True) for zone in dat.spatial.zones).most_common(10)`
(3) `GeoZone.objects().count()`
(4) `Counter(zone.level for zone in GeoZone.objects()).most_common(10)`

# Notes
* We're missing `country:fr` and other national or supranational levels after migration
* It seems to work well for communes

**Review by maudetes — COMMENTED** [review]
Don't forget the changelog (with a migration explanation) please :pray:

**Review by maudetes — COMMENTED** [review]
Nice ! Almost all objects have been moved, but I see exceptions:
* countries (due to code change, except France)
* `country-group:ue:`
* `country-subset:fr:drom`
* `country-subset:fr:dromcom`
* `country-subset:fr:metro`

It would be nice to migrate those, since they seem quite used.
* For countries, I think we can keep ISO Alpha 2 codes in new geozones instead of INSEE's. We can use the ones from [this dataset (filtered on SOUVERAIN)](https://explore.data.gouv.fr/?url=https%3A%2F%2Fwww.data.gouv.fr%2Ffr%2Fdatasets%2Fr%2F2b38f28d-15e7-4f0c-b61d-6ca1d9b1cfa2&SOUVERAIN__exact=O).
* For country subset, I wonder how we can add these if they don't exist in the referential?
* For `ue`, I think it's due to the code that went from `ue` to `eu`. Could you add a special case for that in migration @quaxsze?

**Comment by Pierlou** [comment]
For countries, how do you suggest I handle both sources?
For country subsets, I could change the mapping [here](https://github.com/Pierlou/datagouvfr_data_pipelines/blob/770e8d72773121a4c4290ba3ee16e3a431faca0c/data_processing/geozones/task_functions.py#L51) to include the missing ones (not sure it'll work for all cases, nor if the matching could be perfect)

File: udata/core/spatial/commands.py [hunk 5]
```diff
@@ -1,42 +1,30 @@
...
 log = logging.getLogger(__name__)
 
 
-DEFAULT_GEOZONES_FILE = 'https://github.com/etalab/geozones/releases/download/2019.0/geozones-countries-2019-0-msgpack.tar.xz'
-GEOZONE_FILENAME = 'geozones.tar.xz'
-
-
-def level_ref(level):
-    return DBRef(GeoLevel._get_collection_name(), level)
+DEFAULT_GEOZONES_FILE = 'https://demo.data.gouv.fr/fr/datasets/r/db439811-fd00-4618-a4ba-82b48396df4b'
```
Comment (maudetes) [inline udata/core/spatial/commands.py:26]: I'm confused, why aren't these the files uploaded by @Pierlou?

File: udata/core/spatial/commands.py [hunk 6]
```diff
@@ -218,31 +146,33 @@ def migrate():
...
         for current_zone in dataset.spatial.zones:
             counter['zones'] += 1
-            level, code, validity = geoids.parse(current_zone.id)
-            zone = qs(level=level, code=code).valid_at(validity).first()
+
+            level, code = geoids.parse(current_zone.id)
+            zone = qs(level=level, code=code).first() or qs(code=code).first()
+
             if not zone:
                 log.warning('No match for %s: skipped', current_zone.id)
```
Comment (maudetes) [inline udata/core/spatial/commands.py:161]: When testing today, I still have warnings on `No match for country:fr: skipped`, `country-subset:fr:dromcom`, `country-group:world: skipped`, etc.
Shouldn't this have been fixed?

**Review by maudetes — COMMENTED** [review]
I'm still confused about the latest levels (see my comments).

Could you add a changelog? :pray:

**Review by maudetes — APPROVED** [review]
I think we're all clear here!

Thank you again for the clean. We'll be able to iterate more easily if needed on GeoZones.

**Pull request merged**